scanner/decisions/0001-scannerv4-apis.md

```diff
@@ -0,0 +1,93 @@
+# 0001 - ScannerV4 APIs definition
+
+- **Status:** Accepted
+- **Created:** [2023-05-09 Tue]
+
+## Context
+
+We are transitioning Stackrox Scanner to use [ClairCore](https://github.com/quay/claircore) as its underlying scanning engine. The goal is to align vulnerability scanning results between Quay/Clair and Stackrox. The transition will be gradual, starting with container image scanning. During this period, the existing Scanner (ScannerV2) and the new ClairCore-based Scanner (ScannerV4) will be simultaneously deployed.
+
+The move to ClairCore and the adoption of ScannerV4 bring the need for new API definitions. Also, ClairCore allows for Clair deployment modes that better support the secured cluster use-case (local scanning), horizontal scalability, and high-availability scenarios. The new APIs ideally should align with ClairCore's capabilities while minimizing changes in Central.
+
+## Decision
+
+ScannerV4 APIs will exclusively use gRPC. ScannerV4 APIs are not backward compatible with ScannerV2.
+
+ScannerV4 will offer modes of operation [akin to Clair's deployment models](https://quay.github.io/clair/howto/deployment.html). Each mode will implement different gRPC services, reflecting the underlying separation between ClairCore's [libindex](https://pkg.go.dev/github.com/quay/claircore/libindex#Libindex) and [libvuln](https://pkg.go.dev/github.com/quay/claircore/libvuln#Libvuln). They will be named "indexer" and "matcher". Both modes can be enabled concurrently.
+
+ScannerV4 will use "index reports" and "vulnerability reports" data models, similar to ClairCore's. Both types will link to a "scannable resource" (e.g., a container image) using a "hash_id" managed by the clients. It uniquely identifies the resource's manifest.
+
+ScannerV4 will replace ScannerV2 as the container image scanner in Central. A new image integration for ScannerV4 will be created and replace ScannerV2's integration (type "clarify"), activated by the feature flag: `ROX_SCANNER_V4_ENABLED`. The "clarify" type will continue to be used for orchestrator and node scanning.
```

Comment:
Not sure if this is related to Scanner deployment. This is a decision about how Central will use the APIs, which is related and relevant because it impacts the API design. Even if you want to call the definition of Scanner usage in the Central part of "deployment", I would still be inclined to say we want to design thinking how Central will use the APIs.

Reply:
I am OK with leaving the feature flag out of the decision.